Fix discovery/deletion of iscsi block devices

bswartz opened the pull request:
This PR modifies the iSCSI attach/detatch codepaths in the following
ways:
1) After unmounting a filesystem on an iSCSI block device, always
flush the multipath device mapper entry (if it exists) and delete
all block devices so the kernel forgets about them.
2) When attaching an iSCSI block device, instead of blindly
attempting to scan for the new LUN, first determine if the target
is already logged into, and if not, do the login first. Once every
portal is logged into, the scan is done.
3) Scans are now done for specific devices, instead of the whole
bus. This avoids discovering LUNs that kubelet has no interest in.
4) Additions to the underlying utility interfaces, with new tests
for the new functionality.
5) Some existing code was shifted up or down, to make the new logic
work.
6) A typo in an existing exec call on the attach path was fixed.

Fixes #59946

```release-note
When attaching iSCSI volumes, kubelet now scans only the specific
LUNs being attached, and also deletes them after detaching. This avoids
dangling references to LUNs that no longer exist, which used to be the
cause of random I/O errors/timeouts in kernel logs, slowdowns during
block-device related operations, and very rare cases of data corruption.
```

bswartz commented:
CLA issue should be fixed

bswartz commented:
/assign @saad-ali

krmayankk commented:
/sig storage

humblec commented:
/assign @humblec

rootfs commented:
cc @bmarzins

rootfs commented:
I am not a multipath expert, looping in @bmarzins

redbaron commented on pkg/volume/util/device_util_linux.go:
```diff
@@ -80,3 +83,202 @@ func (handler *deviceHandler) FindSlaveDevicesOnMultipath(dm string) []string {
+			// The list of block devices on the scsi bus will be in a
+			// directory called "target%d:%d:%d".
+			// See drivers/scsi/scsi_scan.c in Linux
+			// We assume the channel/bus and device/controller are always zero for iSCSI
```
is it true for iSCSI offloading cards?

bswartz replied:
I'm trying to determine this. I don't have an offload card but I've ordered some and when I get them installed I'll do experiments.

rootfs commented:
@humblec do you have a multipath setup to test this?

rootfs commented:
/ok-to-test

humblec commented:
@bswartz Thanks.. I am starting the review with a question :) 

` After unmounting a filesystem on an iSCSI block device, always
    flush the multipath device mapper entry (if it exists) and delete
    all block devices so the kernel forgets about them.`

Is it really required to flush the device mapper multipath entries ? afaik, the device mapper will take care of flushing stale device entries after a time stamp/period. I am trying to understand why we need to do it by code , if its required all storage admin has to do this operation manually as soon as an unmount occurs which I think not the case. Also without 'mandatory' flushing are we seeing any functional issues in iscsi workflow ?

redbaron commented:
@humblec , AFAIK multipathd reacts to udev events and therefore can lag behind processing them. IMHO it is best to delete multipath device explicitly to avoid any potential issues and bugs in other parts of the chain.

bswartz commented:
@humblec What redbaron says is true. Also, I was following RedHat's recommended practices from this document:
https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/5/html/online_storage_reconfiguration_guide/removing_devices

humblec commented:
@redbaron @bswartz its definitely good to have, however my attempt was to find out whether its mandatory or any potential issues in absense of it. The guide which was pointed above belongs to RHEL 5, the device mapper and other software stack would have got good amount of changes in latest versions, somehow I was not able to find out this section in RHEL 7 guide here https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html-single/dm_multipath/ or https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/storage_administration_guide/index.

humblec commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -242,12 +263,79 @@ func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) (string, error) {
+	// Build a map of SCSI hosts for each target portal. We will need this to
+	// issue the bus rescans.
+	portalHostMap, err := b.deviceUtil.GetIscsiPortalHostMapForTarget(b.Iqn)
+	if nil != err {
```
`err != nil` looks consistent.

bswartz replied:
Fixed

humblec commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -242,12 +263,79 @@ func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) (string, error) {
+	if nil != err {
+		return "", err
+	}
+	glog.V(6).Infof("AttachDisk portal->host map for %s is %v", b.Iqn, portalHostMap)
```
We can place this on v3 or v4.

bswartz replied:
Fixed

humblec commented on pkg/volume/util/device_util_linux.go:
```diff
@@ -80,3 +83,202 @@ func (handler *deviceHandler) FindSlaveDevicesOnMultipath(dm string) []string {
 }
+
+// GetIscsiPortalHostMapForTarget given a target iqn, find all the scsi hosts logged into
+// that target. Returns a map by target portal to
```
Can you please expand the source code comment on how the portalmap looks like ?

bswartz replied:
Fixed

humblec commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -242,12 +263,79 @@ func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) (string, error) {
-		if err != nil {
-			glog.Errorf("iscsi: failed to rescan session with error: %s (%v)", string(out), err)
+		hostNumber, loggedIn := portalHostMap[tp]
+		if !loggedIn {
```
How are we addressing the race condition here, I mean before we login another routine does login or we assume its already logged in and before performing next operation, it was logged out.

bswartz replied:
I'm not sure the old code addressed that case. When I wrote this I had assumed there was locking at a higher level. I can add proper protection from races here in attach/detach but it will either require a Big Giant Lock or some very complex locking on the individual portals.

redbaron commented:
@humblec I dont think it is necessary  and ideally multipathd would remove device once all paths are deleted. 

Removing device explicitly in the code has following benefits in my view:
 - it is less likely to trigger bugs in interaction between multipathd, udev and kernel. There are so many combination spanning  number of years in development between them, there bound to be differences in behaviour which hard to predict. 
 - it  generates much more visible kubelet log entry should  flush operation fail, for instance if device is still open for whatever reason
 - explicit is better than implicit. it is good to have code for every effect we anticipate to happen. Not everybody (==me) is fluent with iSCSI and things which are obvious once you dive deep in the topic might be surprising, when someone else try to debug another instance of odd behaviour.

humblec commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -242,12 +263,79 @@ func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) (string, error) {
+				lastErr = fmt.Errorf("iscsi: failed to sendtargets to portal %s output: %s, err %v", tp, string(out), err)
+				continue
+			}
+			err = updateISCSINode(b, tp)
```
Does updateISCSIDIscoverydb() or updateISCSINode() can have a inconsistent view of portal map compared to iscsi db ?

bswartz replied:
I didn't write that code -- my changed just moved it. Based on what I know about iscsiadm, it stores some things in its databases, but probably not the SCSI hosts -- those it gets at runtime.

Are you worried about a new bug here, or are you considering an alternative implementation?

redbaron commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -447,6 +547,22 @@ func (util *ISCSIUtil) DetachDisk(c iscsiDiskUnmounter, mntPath string) error {
+		return nil
+	}
+	// if device is no longer used, see if need to logout the target
+	refCount, err := getDevicePrefixRefCount(c.mounter, prefix)
```
not sure if it is in scope of this PR, but IMHO kubelet shouldn't try to logout from iSCSI target. iSCSI session could be initiated at the host level before kubelet starts, logging out would yank iSCSI devices which are needed for host, but not needed by kubelet.

bswartz replied:
This is a reasonable point, but the counter argument is that if kubelet never logs out of sessions, then we will leak iscsi sessions over time. Some storage controllers (Solidfire, for example) create a new session for each PV, so over the course of several weeks of running workloads, hundreds or thousands of defunct iSCSI sessions could accumulate.

It think it's better to document that, if you're using iSCSI for the root FS or some other host service that kubernetes doesn't know about, it must not share the iSCSI session with any storage backend. It would not be hard for an admin to set up multiple iSCSI targets to avoid that type of sharing.

bswartz commented:
/retest

rootfs commented:
@bswartz following up. Some volumes prevent simultaneous setup and teardown. Please find an example [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/cinder/cinder.go#L70)

redbaron commented:
It would be really good  to see it in 1.11. @rootfs , @saad-ali do you think it is possible to add it to the milestone?

bswartz commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -438,6 +544,22 @@ func (util *ISCSIUtil) DetachDisk(c iscsiDiskUnmounter, mntPath string) error {
+		glog.Warningf("iscsi detach disk: failed to delete devices\nError: %v", err)
+		// Fall through -- even if deleting fails, a logout may fix problems
+	}
+	cnt--
```
Thanks, I was too hasty with my rebase and I didn't test it

redbaron commented:
@bswartz , closed? :-/

bswartz commented:
@redbaron Sorry it looks like I keep fat-fingering my git pushes, and I somehow pushed master instead of the bugfix branch. I need to figure out how to reopen this.

bswartz commented:
/reopen

bswartz commented:
/reopen

bswartz commented:
Okay I finally sorted out the branch.

bswartz commented:
/retest

redbaron commented:
1.11 has been released without this fix included :(   I don't even know who to CC to ask to track it for 1.12 :(

redbaron commented:
@mohammedzee1000 , I saw you were adding milestone 1.12 to other tickets, how this issue can be added to a future release to be tracked?  I fear without it it is going to rot forever

jsafrane commented:
@humblec @rootfs, I don't see any serious objections and this PR seems to fix a real problem. Any objections against merging it?

rootfs commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -394,6 +448,67 @@ func globalPDPathOperation(b iscsiDiskMounter) func(iscsiDiskMounter, string, *I
+	filename := fmt.Sprintf("/sys/block/%s/device/delete", deviceName)
+	fd, err := os.OpenFile(filename, os.O_WRONLY, 0)
+	if err != nil {
+		return err
```
if the file doesn't exit, the device is already gone, don't error out here.

bswartz replied:
Good point.

rootfs commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -212,6 +213,26 @@ func (util *ISCSIUtil) loadISCSI(conf *iscsiDisk, mnt string) error {
+	}
+	defer fd.Close()
+
+	scanCmd := fmt.Sprintf("0 0 %d", lunNumber)
```
why `0 0 l`? if the channel and target are not 0, 0, this won't work. https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/5/html/online_storage_reconfiguration_guide/adding_storage-device-or-path

does this function help? https://github.com/NetApp/kubernetes/blob/04cac2212bdfa65ce717cf72242a189d96878352/pkg/volume/fc/fc_util.go#L123

bswartz replied:
The FC scan function you linked is equivalent to the old behavior, and is what causes the bug. Writing 3 hypens "- - -" to the scan file causes the SCSI subsystem to scan the whole bus, finding disks that are not relevant to the current attach operation. The whole point of this new method is that is scans only a single LUN. I can expand the comment to make this more clear.

rootfs commented on pkg/volume/util/device_util_linux.go:
```diff
@@ -80,3 +83,206 @@ func (handler *deviceHandler) FindSlaveDevicesOnMultipath(dm string) []string {
+
+// FindDevicesForLun given an iqn, and lun number, find all the devices
+// corresponding to that LUN.
+func (handler *deviceHandler) FindDevicesForLun(targetIqn string, lun int) ([]string, error) {
```
nit: `FindDevicesForISCSILun` since this is specific for iscsi

bswartz replied:
Will do

rootfs commented on pkg/volume/util/device_util_linux.go:
```diff
@@ -80,3 +83,206 @@ func (handler *deviceHandler) FindSlaveDevicesOnMultipath(dm string) []string {
+	if err != nil {
+		return nil, err
+	}
+	for _, hostDir := range hostDirs {
```
Can [Glob](https://golang.org/pkg/path/filepath/#Glob) help to simply the flow here?

bswartz replied:
I'd have to write it both ways and compare the complexity. I suspect that using globs would make some of the parsing logic more complex, and also require a rewrite of the unit tests.

rootfs commented on pkg/volume/util/device_util_linux.go:
```diff
@@ -80,3 +83,206 @@ func (handler *deviceHandler) FindSlaveDevicesOnMultipath(dm string) []string {
+	if err != nil {
+		return nil, err
+	}
+	for _, hostDir := range hostDirs {
```
same Glob question

rootfs commented:
@bswartz thanks, some questions and nit. 

/approve

rootfs commented:
status/approved-for-milestone

bswartz commented:
@rootfs I think the command is "/status approved-for-milestone"

rootfs commented:
/status approved-for-milestone

bswartz commented:
/retest

bswartz committed:
Avoid deleted iSCSI LUNs in the kernel

This change ensures that iSCSI block devices are deleted after
unmounting, and implements scanning of individual LUNs rather
than scanning the whole iSCSI bus.

In cases where an iSCSI bus is in use by more than one attachment,
detaching used to leave behind phantom block devices, which could
cause I/O errors, long timeouts, or even corruption in the case
when the underlying LUN number was recycled. This change makes
sure to flush references to the block devices after unmounting.

The original iSCSI code scanned the whole target every time a LUN
was attached. On storage controllers that export multiple LUNs on
the same target IQN, this led to a situation where nodes would
see SCSI disks that they weren't supposed to -- possibly dozens or
hundreds of extra SCSI disks. This caused 3 significant problems:

1) The large number of disks wasted resources on the node and
caused a minor drag on performance.
2) The scanning of all the devices caused a huge number of uevents
from the kernel, causing udev to bog down for multiple minutes in
some cases, triggering timeouts and other transient failures.
3) Because Kubernetes was not tracking all the "extra" LUNs that
got discovered, they would not get cleaned up until the last LUN
on a particular target was detached, causing a logout. This led
to significant complications:

In the time window between when a LUN was unintentially scanned,
and when it was removed due to a logout, if it was deleted on the
backend, a phantom reference remained on the node. In the best
case, the phantom LUN would cause I/O errors and timeouts in the
udev system. In the worst case, the backend could reuse the LUN
number for a new volume, and if that new volume were to be
scheduled to a pod with a phantom reference to the old LUN by the
same number, the initiator could get confused and possibly corrupt
data on that volume.

To avoid these problems, the new implementation only scans for
the specific LUN number it expects to see. It's worth noting that
the default behavior of iscsiadm is to automatically scan the
whole bus on login. That behavior can be disabled by setting
node.session.scan = manual
in iscsid.conf, and for the reasons mentioned above, it is
strongly recommended to set that option. This change still works
regardless of the setting in iscsid.conf, and while automatic
scanning will cause some problems, this change doesn't make the
problems any worse, and can make things better in some cases.

bswartz commented:
Squashed 5 commits down to 1

redbaron commented:
/test pull-kubernetes-e2e-gce

rootfs commented:
/lgtm

redbaron commented:
Big thanks to everybody who was involved fixing it
